test/test_ops.py

```diff
@@ -378,7 +378,7 @@ def _create_tensors_with_iou(self, N, iou_thresh):
         boxes[:, 2:] += boxes[:, :2]
         boxes[-1, :] = boxes[0, :]
         x0, y0, x1, y1 = boxes[-1].tolist()
-        boxes[-1, 2] += (x1 - x0) * (1 - iou_thresh) / iou_thresh
+        boxes[-1, 2] += (x1 - x0) * (1 - iou_thresh) / (iou_thresh - 1e-5)
```

Comment:
Although this might fix the tests, I'm not clear where in the current implementation the rounding is making things behave differently, so the "bug" still persists, but it's not tested anymore.

Do you have any insights on where the discrepancy might be?

Reply:
Not precisely, but with this code driving the existing test functions:

```
N = 30

for s in range(50):
    for iou in [0.2, 0.5, 0.8]:
        torch.manual_seed(s)
        boxes, scores = _create_tensors_with_iou(N, iou)
        r_cpu = ops.nms(boxes, scores, iou).tolist()
        r_cuda = ops.nms(boxes.cuda(), scores.cuda(), iou).tolist()
        if r_cpu != r_cuda:
            excess_boxes = "CPU" if set(r_cpu).difference(set(r_cuda)) else "CUDA"
            endbox_iou = ops.box_iou(boxes[0].unsqueeze(0), boxes[N-1].unsqueeze(0))
            print(f'Seed: {s:3}, Excess: {excess_boxes:4}, iou diff: {(iou - endbox_iou[0][0]):+.12f}')
```

I can see that the excess (i.e. non-removed by CPU or CUDA) box in each test appears to depend on whether the difference between the target / threshold IoU value and the constructed 1st/last-box IoU is positive or negative:

```
$ python test_foo.py
Seed:   3, Excess: CUDA, iou diff: -0.000000059605
Seed:   7, Excess: CPU , iou diff: +0.000000000000
Seed:  12, Excess: CUDA, iou diff: -0.000000119209
Seed:  14, Excess: CPU , iou diff: +0.000000000000
Seed:  25, Excess: CUDA, iou diff: -0.000000059605
Seed:  31, Excess: CPU , iou diff: +0.000000000000
Seed:  32, Excess: CPU , iou diff: +0.000000059605
Seed:  35, Excess: CUDA, iou diff: -0.000000029802
Seed:  41, Excess: CPU , iou diff: +0.000000000000
```

The differences between target IoU and constructed IoU are pretty narrow in the failing cases. Searching over 5000 random seeds, I see 1421 failures, with the target/constructed differences ranging from `-1.788139343262e-07` to `1.192092895508e-07`

In the passing cases, the target/constructed IoU differences are generally larger. I see differences ranging from `-8.642673492432e-06`  through `1.275539398193e-05` (with one outlier at `0.002531647682`!).

Reply:
Easier to see what I'm describing there as a histogram of difference between target & constructed IoU vs whether the test passes or shows extra boxes in CPU or CUDA result:

```
Threshold - constructed IoU Pass  CPU CUDA
--------------------------- ---- ---- ----
-0.000008642673492431640625    1    0    0
-0.000005960464477539062500    1    0    0
-0.000005662441253662109375    1    0    0
-0.000004649162292480468750    1    0    0
-0.000003635883331298828125    1    0    0
-0.000002622604370117187500    1    0    0
-0.000002503395080566406250    1    0    0
-0.000002205371856689453125    1    0    0
-0.000001966953277587890625    1    0    0
-0.000001788139343261718750    1    0    0
-0.000001311302185058593750    1    0    0
-0.000001251697540283203125    2    0    0
-0.000001192092895507812500    2    0    0
-0.000001013278961181640625    1    0    0
-0.000000894069671630859375    3    0    0
-0.000000834465026855468750    1    0    0
-0.000000715255737304687500    4    0    0
-0.000000655651092529296875    1    0    0
-0.000000596046447753906250    5    0    0
-0.000000536441802978515625    3    0    0
-0.000000476837158203125000    7    0    0
-0.000000417232513427734375    5    0    0
-0.000000357627868652343750    9    0    0
-0.000000298023223876953125   14    0    0
-0.000000238418579101562500   26    0    0
-0.000000178813934326171875   69    0   20
-0.000000119209289550781250  254    0   92
-0.000000059604644775390625  684    0  298
 0.000000000000000000000000 1423   98    0
 0.000000059604644775390625 1035  159    0
 0.000000119209289550781250  484   40    0
 0.000000178813934326171875  140    0    0
 0.000000238418579101562500   42    0    0
 0.000000298023223876953125   13    0    0
 0.000000357627868652343750    7    0    0
 0.000000417232513427734375    8    0    0
 0.000000476837158203125000    7    0    0
 0.000000536441802978515625    7    0    0
 0.000000596046447753906250    3    0    0
 0.000000655651092529296875    6    0    0
 0.000000715255737304687500    2    0    0
 0.000001013278961181640625    1    0    0
 0.000001072883605957031250    1    0    0
 0.000001132488250732421875    1    0    0
 0.000001192092895507812500    1    0    0
 0.000001370906829833984375    1    0    0
 0.000001430511474609375000    1    0    0
 0.000001490116119384765625    1    0    0
 0.000001549720764160156250    1    0    0
 0.000001728534698486328125    2    0    0
 0.000001966953277587890625    1    0    0
 0.000012755393981933593750    1    0    0
 0.002531647682189941406250    1    0    0
```

Reply:
I would be ok with merging this change (and thus not really exercising the test case I wanted with the original PR introducing this change), but I think it's more trouble than it's worth it.

Reply:
And a concrete failing snippet:

```
$ cat nms.py 
import torch
from torchvision import ops

torch.set_printoptions(precision=12)

iou = 0.8

scores = torch.tensor([0.75, 0.25])
boxes = torch.tensor([[ 0.426352024078369140625000, 10.556936264038085937500000,
                       29.010599136352539062500000, 13.252419471740722656250000],
                      [0, 0, 0, 0]])

boxes[-1, :] = boxes[0, :]
x0, y0, x1, y1 = boxes[-1].tolist()
boxes[-1, 2] += (x1 - x0) * (1 - iou) / iou

print("Box[0]: ", boxes[0])
print("Box[1]: ", boxes[1])
print("IoU target: ", iou, ", IoU actual: ", ops.box_iou(boxes[0].unsqueeze(0), boxes[1].unsqueeze(0)).squeeze(0).item())
print("CPU:  ", ops.nms(boxes, scores, iou))
print("CUDA: ", ops.nms(boxes.cuda(), scores.cuda(), iou))
```

Which gives:

```
$ python nms.py 
Box[0]:  tensor([ 0.426352024078, 10.556936264038, 29.010599136353, 13.252419471741])
Box[1]:  tensor([ 0.426352024078, 10.556936264038, 36.156661987305, 13.252419471741])
IoU target:  0.8 , IoU actual:  0.8000000715255737
CPU:   tensor([0])
CUDA:  tensor([0, 1], device='cuda:0')
```